### DISCUSSION TITLE
Implementing Immediate Stop Orders Post-Market Order Execution

### DISCUSSION THREAD

**Discussion by pythad** [gbeced/basana#23] [Q&A] [answered]
I am currently working on integrating a `PositionManager` component into my trading system, utilizing your library for algorithmic trading. Your library has been instrumental in the development of my trading strategy, and I greatly appreciate the effort you've put into making it accessible and efficient.

Current backtesting exchange implementation executes market orders at the next bar's open, updating balances accordingly. The reasons for this are clear. However, this delays ability to place corresponding stop orders since balances aren't updated until after execution. This make it impossible to use `create_stop_order` in the same `on_trading_signal` execution with `create_market_order`.

Could you recommend a method within Basana to place stop orders immediately or at least as soon as possible after market orders? I am currently thinking about preserving a need for a stop order in the internal state of a PositionManager and placing it inside `PositionManager.on_bar_event` on the next bar. This should work, but I am not sure it this would be the best approach. Maybe you have some other ideas on this.

Appreciate any insights or best practices you can share for integrating this with your library's order management system.

Thanks.

**Comment by gbeced** [comment]
The limitation here is related to the event based nature of the *backtester*. For any order to get filled a BarEvent needs to take place. When you submit an order, Market, Limit, whatever, those are validated and accepted by the exchange, and they remain in that state until the next BarEvent. It is for that reason that the create_market_order + create_stop_order, or any other combination, won't work with the backtesting exchange. Because you need a BarEvent for an order to get filled.
With a real exchange of course that problem doesn't exist.

I'd suggest a couple of things to mitigate this limitation:

1. Use bars with a higher frequency. If you're using daily bars, try hourly bars. This is to get the 2nd order to get filled as close as possible to the first one.
2. Right after creating the market order, [schedule](https://basana.readthedocs.io/en/latest/basana.html#basana.EventDispatcher.schedule) a callback to check if the first order is filled and then submit the stop order. The key here is to schedule the callback right after the next bar:

```
event_dispatcher.schedule(event_dispatcher.now() + datetime.timedelta(minutes=61), check_order)
```

This is assuming you're using hourly bars, so 1 minute after the next bar the check_order coroutine will be called and you can check and submit the stop order. The nice thing about this approach is that will also work with a real exchange. You just don't need to schedule after the next bar, you can schedule a check in 5 minutes.

Hope this helps.

**Comment by gbeced** [comment, marked as answer]
I've addressed this issue at https://github.com/gbeced/basana/commit/5d55772883b8b85510314e0439826165a13f41b2 and it will be available in the next version.
It is currently disabled by default, for backwards compatibility reasons, but the examples have been updated to enable immediate order processing.
